### PR TITLE
Notify the user on failed import for permission denied (and max file size, already integrated in develop)

### PR DESCRIPTION
**Pull request by JohnMcLear**
Details are on the github issue.

**Edit**: this is a follow up to #3864.

### CODE REVIEW

**Comment by muxator** [comment]
This should go in **1.8.3**, but there are conflicts.

**Comment by muxator** [comment]
Conflicts solved.

**Comment by muxator** [comment]
I'll continue in the next days.

**Comment by JohnMcLear** [comment]
@muxator bump :)

File: src/static/js/collab_client.js [hunk 1]
```diff
@@ -159,6 +159,9 @@ function getCollabClient(ace2editor, serverVars, initialUserInfo, options, _pad)
         rev=newRev;
         if (msg.type == "ACCEPT_COMMIT")
         {
+          // useful for knowing if a user has committed content to pad or not
+          clientVars.thisUserHasEditedThisPad = true;
```
Comment (muxator) [inline src/static/js/collab_client.js:163]: Please initialize this variable to `false` somewhere earlier (perhaps at first client connection?)

File: src/node/handler/ImportHandler.js [hunk 2]
```diff
@@ -82,6 +82,12 @@ async function doImport(req, res, padId)
         if (err) {
           console.warn("Uploading Error: " + err.stack);
         }
+
+        // I hate doing indexOf here but I can't see anything to use...
+        if (err.stack.indexOf("maxFileSize") !== -1) {
+          reject("maxFileSize");
+        }
```
Comment (muxator) [inline src/node/handler/ImportHandler.js:89]: Doesn't exist a way of properly bubbling the error?

The server logs are not so good looking when an import is rejected when a file is too big:

```
When the 
[2020-04-21 22:02:53.499] [WARN] console - Uploading Error: Error: maxFileSize exceeded, received 11353 bytes of file data
    at Stream.<anonymous> (<BASE>/src/node_modules/formidable/lib/incoming_form.js:222:19)
    at Stream.emit (events.js:210:5)
    at Stream.EventEmitter.emit (domain.js:475:20)
    at MultipartParser.parser.onPartData (<BASE>/src/node_modules/formidable/lib/incoming_form.js:387:14)
    at callback (<BASE>/src/node_modules/formidable/lib/multipart_parser.js:102:31)
    at dataCallback (<BASE>/src/node_modules/formidable/lib/multipart_parser.js:115:11)
    at MultipartParser.write (<BASE>/src/node_modules/formidable/lib/multipart_parser.js:237:15)
    at IncomingForm.write (<BASE>/src/node_modules/formidable/lib/incoming_form.js:159:34)
    at IncomingMessage.<anonymous> (<BASE>/src/node_modules/formidable/lib/incoming_form.js:125:12)
    at IncomingMessage.emit (events.js:210:5)
```

Comment (JohnMcLear) [reply src/node/handler/ImportHandler.js:89]: No, this error is coming from formidable.

**Comment by JohnMcLear** [comment]
1 & 2 are handled.  3 Does bubble up afaik but we don't suppress the server side msg.

File: src/node/handler/ImportHandler.js (hunk 2 above)
Comment (muxator) [reply src/node/handler/ImportHandler.js:89]: I see. But a way to catch it must exist.

Let's skim on it for now (== forever). Closing.

**Comment by muxator** [comment]
> 1 & 2 are handled

Indeed. Perfect, thanks!

**Comment by muxator** [comment]
There is still one last thing I want to see & I am having a (probably unrelated) nasty crash on pad loading that is blocking me.

File: src/static/js/pad_editbar.js [hunk 3]
```diff
@@ -392,6 +392,38 @@ var padeditbar = (function()
...
+
+                /*
+                 * I'm setting this because it might be useful in the future.
+                 * Also, we don't want to iterate authors every time we click
+                 * import/export
+                 */
+                hasEdits = true;
+              }
+            });
+          }
```
Comment (muxator) [inline src/static/js/pad_editbar.js:417]: This code covers the case in which we open a pad (or hit F5) where we have already contributed.

Shoudn't be the server to already send us the correct value for `thisUserHasEditedThisPad`? He is the one setting `historicalAuthorData`, after all.

**Comment by muxator** [comment]
I have split the functionalities in this PR. Now they are independent.

1. the first one implements the error message when the imported file is too big. This works very well and does not need any modifications;
2. the second one disables/enables the import UI when the user contributes to the pad at least once. This requires a modification moving some logic from the client to the server side.

I am going to import 1, which is ready, directly on the `develop` branch.

This PR now only deals with 2 (import permission), which is the only one needing fixes.

**Comment by JohnMcLear** [comment]
RE: 

> the second one disables/enables the import UI when the user contributes to the pad at least once. This requires a modification moving some logic from the client to the server side.

The server does check to ensure the client exists..  The client side code is just to provide nice UX.

I don't see much benefit moving the logic to the server?

**Comment by muxator** [comment]
> I don't see much benefit moving the logic to the server?

It's the server that sends the client variables to the client on first connection.

`thisUserHasEditedThisPad` is a new variable we are introducing in this PR, whose value on first connection was `undefined` (before my comment) and now `false` (after my request of initializing it).

I'd say it's safe to expect that something that is called `thisUserHasEditedThisPad` lives up to its name. It's the server that creates it, he knows what to do.

Moving the logic on the server and removing the client side code guarantees an invariant, that otherwise we are introducing and breaking at the same time.


If we do not do this, we are re-enacting one of the main ways of giving life to bugs.

File: src/static/js/pad_editbar.js (hunk 3 above)
Comment (JohnMcLear) [reply src/static/js/pad_editbar.js:417]: I don't see the gain.  Either the server sends a boolean or we iterate the object and set the boolean.  Either way the boolean is set.

**Comment by JohnMcLear** [comment]
Okay noted RE shifting code to server.  Will look to see if this is workable within time constraints.

**Comment by JohnMcLear** [comment]
Fix is in, give that a go :)

**Comment by muxator** [comment]
> Fix is in, give that a go :)

It is **perfect** now! Kudos!

I've just squashed the commits and moved a comment around.

**Comment by JohnMcLear** [comment]
Tnx, sorry for push back you were right to make me rework it

**Pull request merged**